tests/host_tools/logging.py

```diff
@@ -2,35 +2,43 @@
 # SPDX-License-Identifier: Apache-2.0
 """Utilities for testing the logging system (metrics, common logs)."""
 
-import fcntl
 import os
 import sys
 
 from queue import Queue
-from subprocess import run
 from threading import Thread
 
 
 class Fifo:
     """Facility for creating and working with named pipes (FIFOs)."""
 
     path = None
+    fifo = None
 
     def __init__(self, path):
         """Create a new named pipe."""
         if os.path.exists(path):
             raise FileExistsError("Named pipe {} already exists.".format(path))
-        cmd = 'mkfifo ' + path
-        run(cmd, shell=True, check=True)
+        os.mkfifo(path)
+        # Open the FIFO and set O_NONBLOCK flag.
+        fd = os.open(path, os.O_RDONLY | os.O_NONBLOCK)
+        self.fifo = os.fdopen(fd, "r")
         self.path = path
 
     def sequential_reader(self, max_lines):
-        """Return up to `max_lines` lines from fifo `fifo_index`.
+        """Return up to `max_lines` lines from a non blocking fifo.
 
         :return: A list containing the read lines.
         """
-        fifo = self._open_nonblocking()
-        return fifo.readlines()[:max_lines]
+        return self.fifo.readlines()[:max_lines]
+
+    def sequential_blocking_reader(self, max_lines):
+        """Return up to `max_lines` lines from fifo which blocks.
+
+        :return: A list containing the read lines.
+        """
+        blocking_fifo = open(self.path, "r")
```

Comment:
Why open a new one instead of `fcntl`'ing the existing fd?

Reply:
Actually, why not specify in the constructor if you want the fifo to be blocking or not, and use a `self.blocking` attribute to control the behavior?

Reply:
Yes, specifying the blocking behavior in the constructor makes sense. I will add also a method for changing the blocking behavior. E.g. opening a pipe in blocking mode, use it, then deciding that we want to use it later on with non blocking behavior, or vice versa. WDYT?